artiq/applets/big_number.py

```diff
@@ -26,47 +27,71 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = None
+        self.metadata = {}
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
         self.edit_widget.editCancelled.connect(self.cancel_edit)
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
         font = QtGui.QFont()
         font.setPointSize(60)
+        self.display_widget.setFont(font)
         self.edit_widget.setFont(font)
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            scale = scale_from_metadata(self.metadata)
+            value = self.value/scale
+            unit = self.metadata.get("unit", "")
+            self.edit_widget.setText(f"{value} {unit}".strip())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
+        text = self.edit_widget.text().split() or "---"
+        try:
+            value = float(text[0])
+        except ValueError:
+            self.display_widget.setText("---")
+            self.cancel_edit()
+            return
+        unit = text[1] if len(text) > 1 else ""
+        scale = scale_from_metadata({"unit": unit})
+        value = value*scale
+        precision = self.metadata.get("precision")
+        self.ctl.set_dataset(self.dataset_name, value, unit=unit,
+                             precision=precision)
+        self.setCurrentWidget(self.display_widget)
 
     def cancel_edit(self):
```

Comment:
Considering this is a one-liner, I propose either: A) rename it to something like `end_edit` (or something that does not imply cancel, can find a better name). Then use this function in place of all `self.setCurrentWidget(self.display_widget)` or B) remove this function and in the editCancelled pass a lambda instead.

Reply:
Updated with option B